core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -963,8 +1007,15 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     } replayBus.addListener(listener)
 
     try {
+      val eventLogFiles = reader.listEventLogFiles
+      val newEventLogFiles = if (compactible.contains(true)) {
+        logInfo(s"Compacting ${reader.rootPath}...")
+        fileCompactor.compact(eventLogFiles)
```

Comment:
Just to follow up on my previous comment, I think tying this to the UI rebuild code is not a great idea, since it means that logs may never be compacted. So perhaps that PR that allows tasks to run in parallel in the SHS really can help here (then you can compact during the `checkLogs()` tasks).

Reply:
> I think tying this to the UI rebuild code is not a great idea, since it means that logs may never be compacted.

You're right. I wanted to simplify the thing first given it would couple with thread-safety while the code change already goes beyond 2000+ lines, but I have been also feeling that it should be fixed.

> So perhaps that PR that allows tasks to run in parallel in the SHS really can help here (then you can compact during the checkLogs() tasks).

Yes but given #25797 is still in reviewing I'm not sure how I can leverage it. Pick the PR up and try to finalize that first?

Reply:
I just moved the compaction into checkLogs() tasks - once #25797 is merged I'll rebase and fix if there's any conflict.